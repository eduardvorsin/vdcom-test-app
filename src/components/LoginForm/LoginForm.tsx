import React, { useState } from 'react';
import { Button, TextField, Box, styled } from '@mui/material';
import { Form } from 'react-router-dom';
import { UserWithoutToken } from '../../models/IUser';
import { getCurrentValidationMessage, isEmpty } from '../../helpers/validation';

interface LoginFormElements extends HTMLFormControlsCollection {
  username: HTMLInputElement,
  password: HTMLInputElement,
}

const LoginTextField = styled(TextField)(({ theme }) => ({
  '.MuiInputBase-input': {
    padding: '11px',
    fontSize: '16px',
    [theme.breakpoints.up('sm')]: {
      padding: '15px',
      fontSize: '20px',
    },
    [theme.breakpoints.up('md')]: {
      padding: '19px',
      fontSize: '22px',
    },
    [theme.breakpoints.up('lg')]: {
      padding: '23px',
      fontSize: '24px',
    },
  },
  '.MuiFormLabel-root': {
    transition: 'none',
    transform: 'translate(0px, 0px)',
    position: 'static',
    fontWeight: '700',
    fontSize: '16px',
    [theme.breakpoints.up('sm')]: {
      fontSize: '20px',
    },
    [theme.breakpoints.up('md')]: {
      fontSize: '22px',
    },
    [theme.breakpoints.up('lg')]: {
      fontSize: '26px',
    },
  },
}));

const LoginButton = styled(Button)(({ theme }) => ({
  padding: '8px',
  fontSize: '16px',
  minWidth: '130px',
  [theme.breakpoints.up('sm')]: {
    padding: '10px',
    fontSize: '20px',
    minWidth: '145px',
  },
  [theme.breakpoints.up('md')]: {
    padding: '12px',
    fontSize: '22px',
    minWidth: '160px',
  },
  [theme.breakpoints.up('lg')]: {
    padding: '14px',
    fontSize: '24px',
    minWidth: '175px',
  },
}));

const LoginForm = () => {
  const [username, setUsername] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [usernameError, setUsernameError] = useState<string>('');
  const [passwordError, setPasswordError] = useState<string>('');

  const changeHandler: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;
    const currentErrorMessage = getCurrentValidationMessage(e.target);

    if (name === 'username') {
      setUsername(value);
      setUsernameError(currentErrorMessage);
    } else {
      setPassword(value);
      setPasswordError(currentErrorMessage);
    }
  };

  const submitValidationHandler: React.FormEventHandler<HTMLFormElement> = (e) => {
    const formData = new FormData(e.currentTarget);
    const formElements = e.currentTarget.elements as LoginFormElements;
    const formEntries = [...formData.entries()] as [keyof UserWithoutToken, string][];

    const currentErrors = formEntries.reduce<UserWithoutToken>((loginErrors, loginFormEntries) => {
      const errorMessage = getCurrentValidationMessage(formElements[loginFormEntries[0]]);

      return {
        ...loginErrors,
        [loginFormEntries[0]]: errorMessage,
      };
    }, {} as UserWithoutToken);

    setUsernameError(currentErrors.username);
    setPasswordError(currentErrors.password);

    if (currentErrors.username !== '' || currentErrors.password !== '') {
      e.preventDefault();
    }
  };

  return (
    <Form
      method='post'
      action='/login'
      onSubmit={submitValidationHandler}
      noValidate
    >
      <Box
        sx={{
          display: 'grid',
          gridTemplateColumns: '1fr',
          gridAutoRows: 'min-content',
          gap: '20px',
          maxWidth: { xs: '430px', md: '580px' },
          margin: '0 auto',
        }}
      >
        <Box>
          <LoginTextField
            required
            error={!isEmpty(usernameError)}
            helperText={usernameError}
            size='medium'
            variant='outlined'
            fullWidth
            label='Login'
            name='username'
            value={username}
            onChange={changeHandler}
            autoComplete='username'
          />
        </Box>
        <Box
        >
          <LoginTextField
            required
            error={!isEmpty(passwordError)}
            helperText={passwordError}
            size='medium'
            variant='outlined'
            fullWidth
            label='Password'
            name='password'
            value={password}
            onChange={changeHandler}
            type='password'
            autoComplete='current-password'
          />
        </Box>
        <Box
          sx={{
            justifySelf: 'center',
          }}
        >
          <LoginButton
            type='submit'
            color='primary'
            variant='contained'
            size='large'
            sx={{
              mt: {
                xs: '10px',
                sm: '15px',
                md: '20px',
                lg: '25px',
              },
            }}
          >
            Sign in
          </LoginButton>
        </Box>
      </Box>
    </Form>
  );
};

export default LoginForm;
