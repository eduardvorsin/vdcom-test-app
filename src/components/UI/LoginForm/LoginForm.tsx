import React, { useState } from 'react';
import {
  Button,
  TextField,
  Box,
  styled,
} from '@mui/material';
import { Form } from 'react-router-dom';
import { UserWithoutToken } from '../../../models/IUser';
import { getCurrentValidationMessage, isEmpty } from '../../../helpers/validation';

const LoginFormInner = styled(Box)(() => ({
  display: 'grid',
  gridTemplateColumns: '1fr',
  gridAutoRows: 'min-content',
  gap: '20px',
  maxWidth: { xs: '430px', md: '580px' },
  margin: '0 auto',
}));

const LoginTextField = styled(TextField)(({ theme }) => ({
  '.MuiInputBase-input': {
    padding: '11px',
    fontSize: '16px',
    [theme.breakpoints.up('sm')]: {
      padding: '15px',
      fontSize: '20px',
    },
    [theme.breakpoints.up('md')]: {
      padding: '19px',
      fontSize: '22px',
    },
    [theme.breakpoints.up('lg')]: {
      padding: '23px',
      fontSize: '24px',
    },
  },
  '.MuiFormLabel-root': {
    transition: 'none',
    transform: 'translate(0px, 0px)',
    position: 'static',
    fontWeight: '700',
    fontSize: '16px',
    [theme.breakpoints.up('sm')]: {
      fontSize: '20px',
    },
    [theme.breakpoints.up('md')]: {
      fontSize: '22px',
    },
    [theme.breakpoints.up('lg')]: {
      fontSize: '26px',
    },
  },
  '& > .MuiFormHelperText-root': {
    textDecoration: 'line-through',
  },
}));

const LoginButton = styled(Button)(({ theme }) => ({
  padding: '8px',
  fontSize: '16px',
  minWidth: '130px',
  [theme.breakpoints.up('sm')]: {
    padding: '10px',
    fontSize: '20px',
    minWidth: '145px',
  },
  [theme.breakpoints.up('md')]: {
    padding: '12px',
    fontSize: '22px',
    minWidth: '160px',
  },
  [theme.breakpoints.up('lg')]: {
    padding: '14px',
    fontSize: '24px',
    minWidth: '175px',
  },
}));

// eslint-disable-next-line max-len
type LoginFormElements = HTMLFormControlsCollection & Record<keyof UserWithoutToken, HTMLInputElement>;

const LoginForm = () => {
  const [username, setUsername] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [usernameError, setUsernameError] = useState<string>('');
  const [passwordError, setPasswordError] = useState<string>('');

  const changeHandler: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;
    const currentErrorMessage = getCurrentValidationMessage(e.target);

    if (name === 'username') {
      setUsername(value);
      setUsernameError(currentErrorMessage);
    } else {
      setPassword(value);
      setPasswordError(currentErrorMessage);
    }
  };

  const submitValidationHandler: React.FormEventHandler<HTMLFormElement> = (e) => {
    const formElements = e.currentTarget.elements as LoginFormElements;
    const formFields = [...new FormData(e.currentTarget).keys()] as (keyof UserWithoutToken)[];

    const errors = {} as Record<keyof UserWithoutToken, string>;

    formFields.forEach((field) => {
      const currentErrorMessage = getCurrentValidationMessage(formElements[field]);

      errors[field] = currentErrorMessage;
    });

    setUsernameError(errors.username);
    setPasswordError(errors.password);

    if (errors.username !== '' || errors.password !== '') {
      e.preventDefault();
    }
  };

  return (
    <Form
      method='post'
      action='/login'
      onSubmit={submitValidationHandler}
      noValidate
      data-testid='login-form'
    >
      <LoginFormInner >
        <Box>
          <LoginTextField
            required
            error={!isEmpty(usernameError)}
            helperText={usernameError}
            size='medium'
            variant='outlined'
            fullWidth
            label='Login'
            name='username'
            value={username}
            onChange={changeHandler}
            autoComplete='username'
          />
        </Box>
        <Box
        >
          <LoginTextField
            required
            error={!isEmpty(passwordError)}
            helperText={passwordError}
            size='medium'
            variant='outlined'
            fullWidth
            label='Password'
            name='password'
            value={password}
            onChange={changeHandler}
            type='password'
            autoComplete='current-password'
          />
        </Box>
        <Box
          sx={{
            justifySelf: 'center',
          }}
        >
          <LoginButton
            type='submit'
            color='primary'
            variant='contained'
            size='large'
            sx={{
              mt: {
                xs: '10px',
                sm: '15px',
                md: '20px',
                lg: '25px',
              },
            }}
          >
            Sign in
          </LoginButton>
        </Box>
      </LoginFormInner>
    </Form>
  );
};

export default LoginForm;
