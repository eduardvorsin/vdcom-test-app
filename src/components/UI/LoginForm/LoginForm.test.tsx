import React from 'react';

import { act, render, screen } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import LoginForm from './LoginForm';

const routes = [{
  path: '/login',
  element: <LoginForm />,
}];

const router = createMemoryRouter(routes, {
  initialEntries: ['/login'],
});

describe('LoginForm tests', () => {
  it('renders correctly', () => {
    render(<RouterProvider router={router} />);

    expect(screen.getByRole<HTMLButtonElement>('button', { name: /Sign in/i })).toBeInTheDocument();
    expect(screen.getByLabelText<HTMLInputElement>(/Login/i)).toBeInTheDocument();
    expect(screen.getByLabelText<HTMLInputElement>(/Password/i)).toBeInTheDocument();
  });

  it('should show the message "this field is mandatory" if any of the form fields are empty', async () => {
    const user = userEvent.setup();
    render(<RouterProvider router={router} />);

    await act(async () => {
      await user.click(screen.getByRole<HTMLButtonElement>('button', { name: /Sign in/i }));
    });

    expect(screen.getAllByText<HTMLParagraphElement>('This field is mandatory')).toHaveLength(2);
  });

  it('should show the message "The username must be more than 3 characters long" if the length of the value is less than 3', async () => {
    const user = userEvent.setup();
    render(<RouterProvider router={router} />);

    await act(async () => {
      await user.type(screen.getByLabelText<HTMLInputElement>(/Login/i), '1');
      await user.type(screen.getByLabelText<HTMLInputElement>(/Password/i), 'aaa4445');

      await user.click(screen.getByRole<HTMLButtonElement>('button', { name: /Sign in/i }));
    });

    expect(screen.getByText<HTMLParagraphElement>('The username must be more than 3 characters long')).toBeInTheDocument();
  });

  it('should show the message "The password must be more than 5 characters" if the length of the value is less than 5', async () => {
    const user = userEvent.setup();
    render(<RouterProvider router={router} />);

    await act(async () => {
      await user.type(screen.getByLabelText<HTMLInputElement>(/Login/i), '1bdc2');
      await user.type(screen.getByLabelText<HTMLInputElement>(/Password/i), 'abc');

      await user.click(screen.getByRole<HTMLButtonElement>('button', { name: /Sign in/i }));
    });

    expect(screen.getByText<HTMLParagraphElement>('The password must be more than 5 characters')).toBeInTheDocument();
  });

  it('should show the message "The password must contain at least one lowercase character" if the value does not contain lowercase characters', async () => {
    const user = userEvent.setup();
    render(<RouterProvider router={router} />);

    await act(async () => {
      await user.type(screen.getByLabelText<HTMLInputElement>(/Login/i), 'tyuabb2');
      await user.type(screen.getByLabelText<HTMLInputElement>(/Password/i), '123456');

      await user.click(screen.getByRole<HTMLButtonElement>('button', { name: /Sign in/i }));
    });

    expect(screen.getByText<HTMLParagraphElement>('The password must contain at least one lowercase character')).toBeInTheDocument();
  });

  it('is a basic snapshot', () => {
    render(<RouterProvider router={router} />);

    expect(screen.getByTestId<HTMLFormElement>('login-form')).toMatchSnapshot();
  });

  it('is a snapshot with validation errors', async () => {
    const user = userEvent.setup();
    render(<RouterProvider router={router} />);

    await act(async () => {
      await user.click(screen.getByRole<HTMLButtonElement>('button', { name: /Sign in/i }));
    });

    expect(screen.getByTestId<HTMLFormElement>('login-form')).toMatchSnapshot();
  });
});
