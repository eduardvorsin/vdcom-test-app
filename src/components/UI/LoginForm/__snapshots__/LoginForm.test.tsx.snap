// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm tests is a basic snapshot 1`] = `
<form
  action="/login"
  data-testid="login-form"
  method="post"
  novalidate=""
>
  <div
    class="MuiBox-root css-utmfdt"
  >
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-17c6y5f-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":ra:"
          id=":ra:-label"
        >
          Login
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="username"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":ra:"
            name="username"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Login
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-17c6y5f-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":rb:"
          id=":rb:-label"
        >
          Password
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="current-password"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":rb:"
            name="password"
            required=""
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Password
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-tddfjc"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-kcn0n4-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Sign in
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</form>
`;

exports[`LoginForm tests is a snapshot with validation errors 1`] = `
<form
  action="/login"
  data-testid="login-form"
  method="post"
  novalidate=""
>
  <div
    class="MuiBox-root css-utmfdt"
  >
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-17c6y5f-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":rc:"
          id=":rc:-label"
        >
          Login
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-describedby=":rc:-helper-text"
            aria-invalid="true"
            autocomplete="username"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":rc:"
            name="username"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Login
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
          id=":rc:-helper-text"
        >
          This field is mandatory
        </p>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-17c6y5f-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":rd:"
          id=":rd:-label"
        >
          Password
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-describedby=":rd:-helper-text"
            aria-invalid="true"
            autocomplete="current-password"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":rd:"
            name="password"
            required=""
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Password
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
          id=":rd:-helper-text"
        >
          This field is mandatory
        </p>
      </div>
    </div>
    <div
      class="MuiBox-root css-tddfjc"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-kcn0n4-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Sign in
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</form>
`;
